src/loader/riscv64/pe/mod.rs

```diff
@@ -0,0 +1,238 @@
+
+use super::super::{Error as KernelLoaderError, KernelLoader, KernelLoaderResult, Result};
+
+/// RISC-V Image (PE) format support
+pub struct PE;
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for riscv64_image_header {}
+
+#[derive(Debug, PartialEq, Eq)]
+/// PE kernel loader errors.
+pub enum Error {
+    /// Unable to seek to Image end.
+    SeekImageEnd,
+    /// Unable to seek to Image header.
+    SeekImageHeader,
+    /// Unable to seek to DTB start.
+    SeekDtbStart,
+    /// Unable to seek to DTB end.
+    SeekDtbEnd,
+    /// Device tree binary too big.
+    DtbTooBig,
+    /// Unable to read kernel image.
+    ReadKernelImage,
+    /// Unable to read Image header.
+    ReadImageHeader,
+    /// Unable to read DTB image
+    ReadDtbImage,
+    /// Invalid Image binary.
+    InvalidImage,
+    /// Invalid Image magic number.
+    InvalidImageMagicNumber,
+    /// Invalid base address alignment
+    InvalidBaseAddrAlignment,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let desc = match self {
+            Error::SeekImageEnd => "unable to seek Image end",
+            Error::SeekImageHeader => "unable to seek Image header",
+            Error::ReadImageHeader => "unable to read Image header",
+            Error::ReadDtbImage => "unable to read DTB image",
+            Error::SeekDtbStart => "unable to seek DTB start",
+            Error::SeekDtbEnd => "unable to seek DTB end",
+            Error::InvalidImage => "invalid Image",
+            Error::InvalidImageMagicNumber => "invalid Image magic number",
+            Error::DtbTooBig => "device tree image too big",
+            Error::ReadKernelImage => "unable to read kernel image",
+            Error::InvalidBaseAddrAlignment => "base address not aligned to 2 MB",
+        };
+
+        write!(f, "PE Kernel Loader: {}", desc)
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Default)]
+// See kernel doc Documentation/arch/riscv/boot-image-header.rst
+// All these fields should be little endian.
+struct riscv64_image_header {
+    code0: u32,
+    code1: u32,
+    text_offset: u64,
+    image_size: u64,
+    flags: u64,
+    version: u32,
+    res1: u32,
+    res2: u64,
+    magic: u64,
+    magic2: u32,
+    res3: u32,
+}
+
+impl KernelLoader for PE {
+    /// Loads a PE Image into guest memory.
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem` - The guest memory where the kernel image is loaded.
+    /// * `kernel_offset` - 2MB-aligned base addres in guest memory at which to load the kernel.
+    /// * `kernel_image` - Input Image format kernel image.
+    /// * `highmem_start_address` - ignored on RISC-V.
+    ///
+    /// # Returns
+    /// * KernelLoaderResult
+    fn load<F, M: GuestMemory>(
```

Comment:
This is also very similar to the existing aarch64 code. Can we reuse part of it somehow?